components/encryption/src/io.rs

```diff
@@ -17,73 +17,303 @@ use file_system::File;
         key: &[u8],
-    ) -> Result<(EncrypterReader<R>, Iv)> {
-        let (crypter, iv) = CrypterReader::new(reader, method, key, Mode::Encrypt, None)?;
-        Ok((EncrypterReader(crypter), iv))
+        iv: Iv,
+    ) -> Result<EncryptReader<R>> {
+        Ok(EncryptReader(CrypterReader::new(
+            reader,
+            method,
+            key,
+            Mode::Encrypt,
+            iv,
+        )?))
     }
 }
 
-impl<R: Read> Read for EncrypterReader<R> {
+impl<R: Read> Read for EncryptReader<R> {
     fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
         self.0.read(buf)
     }
 }
 
-impl<R: Seek> Seek for EncrypterReader<R> {
+impl<R: Seek> Seek for EncryptReader<R> {
     fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
         self.0.seek(pos)
     }
 }
 
-impl<R: AsyncRead + Unpin> AsyncRead for EncrypterReader<R> {
+impl<R: AsyncRead + Unpin> AsyncRead for EncryptReader<R> {
     fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<IoResult<usize>> {
         unsafe { self.map_unchecked_mut(|r| &mut r.0) }.poll_read(cx, buf)
     }
 }
 
 /// Decrypt content as data being read.
-pub struct DecrypterReader<R>(CrypterReader<R>);
+pub struct DecryptReader<R>(CrypterReader<R>);
 
-impl<R> DecrypterReader<R> {
+impl<R> DecryptReader<R> {
     pub fn new(
         reader: R,
         method: EncryptionMethod,
         key: &[u8],
         iv: Iv,
-    ) -> Result<DecrypterReader<R>> {
-        let (crypter, _) = CrypterReader::new(reader, method, key, Mode::Decrypt, Some(iv))?;
-        Ok(DecrypterReader(crypter))
+    ) -> Result<DecryptReader<R>> {
+        Ok(DecryptReader(CrypterReader::new(
+            reader,
+            method,
+            key,
+            Mode::Decrypt,
+            iv,
+        )?))
     }
 }
 
-impl<R: Read> Read for DecrypterReader<R> {
+impl<R: Read> Read for DecryptReader<R> {
     fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
         self.0.read(buf)
     }
 }
 
-impl<R: Seek> Seek for DecrypterReader<R> {
+impl<R: Seek> Seek for DecryptReader<R> {
     fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
         self.0.seek(pos)
     }
 }
 
-impl<R: AsyncRead + Unpin> AsyncRead for DecrypterReader<R> {
+impl<R: AsyncRead + Unpin> AsyncRead for DecryptReader<R> {
     fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<IoResult<usize>> {
         unsafe { self.map_unchecked_mut(|r| &mut r.0) }.poll_read(cx, buf)
     }
 }
 
+/// Encrypt content as data being written.
+pub struct EncryptWriter<W>(CrypterWriter<W>);
+
+impl<W> EncryptWriter<W> {
+    pub fn new(
+        writer: W,
+        method: EncryptionMethod,
+        key: &[u8],
+        iv: Iv,
+    ) -> Result<EncryptWriter<W>> {
+        Ok(EncryptWriter(CrypterWriter::new(
+            writer,
+            method,
+            key,
+            Mode::Encrypt,
+            iv,
+        )?))
+    }
+
+    pub fn finalize(self) -> IoResult<W> {
```

Comment:
why need to return the inner writer

Reply:
should we call flush explicitly?

Reply:
It's just an idiomatic to deal with chained streams, e.g. https://docs.rs/lz4/1.15.131/lz4/struct.Encoder.html#method.finish or https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner.
This function usually only guarantees bytes are written, no need to be flushed.